wallet-new/test/unit/UTxO/BlockGen.hs

```diff
@@ -0,0 +1,245 @@
+        . fmap (OldestFirst . reverse)
+        . zipFees ks
+
+-- | Given an initial bootstrap 'Transaction', this function will
+-- initialize the generator and run the action provided.
+runBlockGen
+    :: Hash h Addr
+    => Transaction h Addr
+    -> BlockGen h a
+    -> Gen a
+runBlockGen = runBlockGenWith identity
+
+-- | Given an initial bootstrap 'Transaction' and a function to customize
+-- the other settings in the 'BlockGenCtx', this function will initialize
+-- the generator and run the action provided.
+runBlockGenWith
+    :: Hash h Addr
+    => (BlockGenCtx h -> BlockGenCtx h)
+    -> Transaction h Addr
+    -> BlockGen h a
+    -> Gen a
+runBlockGenWith settings boot (BlockGen m) =
+    evalStateT m (settings (initializeCtx boot))
+
+-- | Create an initial context from the boot transaction.
+initializeCtx :: Hash h Addr => Transaction h Addr -> BlockGenCtx h
+initializeCtx boot@Transaction{..} = BlockGenCtx {..}
+  where
+    _bgcAddressesToBalances =
+        foldl' outToValue mempty trOuts
+    outToValue acc (Output {..}) =
+        Map.insertWith (+) outAddr outVal acc
+    _bgcFreshHash = 0
+    _bgcCurrentBlockchain = ledgerSingleton boot
+
+-- | Lift a 'Gen' action into the 'BlockGen' monad.
+liftGen :: Gen a -> BlockGen h a
+liftGen = BlockGen . lift
+
+-- | Provide a fresh hash value for a transaction.
+freshHash :: BlockGen h Int
+freshHash = do
+    i <- use bgcFreshHash
+    bgcFreshHash += 1
+    pure i
+
+-- | Select a random address from the current state.
+selectFromAddr :: BlockGen h Addr
+selectFromAddr =
+    liftGen . elements =<< uses bgcAddressesToBalances (toList . Map.keysSet)
+
+-- | Select a random address from the current state that is not the
+-- provided address.
+selectToAddress :: Addr -> BlockGen h Addr
+selectToAddress addr = do
+    addrs <- uses bgcAddressesToBalances (toList . Set.delete addr . Map.keysSet)
+    liftGen (elements (filter (not . isAvvmAddr) addrs))
+
+-- | Returns true if the 'addrActorIx' is the 'IxAvvm' constructor.
+isAvvmAddr :: Addr -> Bool
+isAvvmAddr addr =
+    case addrActorIx addr of
+        IxAvvm _ -> True
+        _ -> False
+
+-- | Select a random address from the current set of addresses, and
+-- a random value between 0 and that address's total value minus the
+-- possible fee amount. The 'Value' that is returned is not deducted from
+-- the 'Addr's balance.
+selectFromAddrAndAmount :: BlockGen h (Addr, Value)
+selectFromAddrAndAmount = do
+    addrs <- uses bgcAddressesToBalances Map.toList
+    (addr, amount) <- liftGen $ elements addrs
+    value <- liftGen $ choose (1, amount `safeSubtract` maxFee)
+    if value < 1
+        then localBlockGen
+            (bgcAddressesToBalances . at addr .~ Nothing)
+            selectFromAddrAndAmount
+        else pure (addr, value)
+
+maxFee :: Num a => a
+maxFee = 180000
+
+-- | Run a 'BlockGen' action with a modified context. Changes to the state
+-- from the inner action are discarded.
```

Comment:
That I use this function makes me want to give a `MonadReader` instance to the `BlockGen` stack; then programming against `MonadReader (BlockGenCtx h)` would forbid you to run any state-changing functions inside of this. It'd need a somewhat gnarly constraint (`MonadReader (BlockGenCtx h) t, MonadTrans t) => t Gen a` maybe?) without refactoring the types a bit more.

 I recall reading a document in Cardano that strongly advised for programming against concrete monads instead of `mtl`-style constraints, though, so I kept it like this.